### ISSUE TITLE
Warning message for deprecated APIs for some resources show incorrect Kubernetes version 

### ISSUE THREAD

**Issue by hickeyma** [kubernetes/kubernetes#94933]
**What happened**:

When calling API https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/deprecation/deprecation.go#L101, it returns incorrect message for some resources. For example:

```console
extensions/v1beta1 Deployment is deprecated in v1.8+, unavailable in v1.18+; use apps/v1 Deployment
```

The above API is defined in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/extensions/v1beta1/types.go#L80. It would seem that other deprecated and removed versions maybe incorrectly defined in the annotations as well.

**What you expected to happen**:

For example above, I would have expected:

```console
extensions/v1beta1 Deployment is deprecated in v1.9+, unavailable in v1.16+; use apps/v1 Deployment
```

This is because the API is removed in 1.16. This applies to other resources and API versions as well.

**How to reproduce it (as minimally and precisely as possible)**:

- Try to deploy a Deployment object with `apiVersion: extensions/v1beta1` in Kubernetes 1.16. It should fail to deploy.
- https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/ provide a good description of API versions

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.19
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): 
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by hickeyma** [comment]
/sig @kubernetes/sig-apps-misc @kubernetes/sig-api-machinery-misc

**Comment by liggitt** [comment]
that message is actually correct... that stopped being served by default in 1.16 but was completely disabled in 1.18+

for reference:
* https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.16.md#deprecations-and-removals
   > The ability to serve these resources will be completely removed in v1.18.
* https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.18.md#kube-apiserver-1
   > the following deprecated APIs can no longer be served ...

also, for that one, it's a bit academic, since the warning wouldn't be sent prior to 1.19, and the endpoints for that type are gone in 1.18+ :-)

**Comment by hickeyma** [comment]
> that message is actually correct... that stopped being served by default in 1.16 but was completely disabled in 1.18+

It may have become disabled in 1.18 but in 1.16 you are unable to use the resource again. In Helm, we have users that deploy objects in a Kubernetes version when they are still valid. The user may one day upgrade the Kubernetes version which now makes the object invalid or removed. 

> also, for that one, it's a bit academic, since the warning wouldn't be sent prior to 1.19, and the endpoints for that type are gone in 1.18+

Helm provides linting capability to help users identify and potential issues with their charts. It is run client side and does not go to the Kubernetes server. These charts contain the manifests to deploy Kubernetes resources.  Users can potentially have Kubernetes objects that have been removed and  they maybe unaware of this. We would like to be able to provide them with a message that informs them of the correct version when deprecated and removed. Here is the PR for it: https://github.com/helm/helm/pull/8608

**Comment by joejulian** [comment]
Maybe I'm misreading the goal, but isn't the intent of https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings to report when the resource is/will-be no longer served?

**Comment by liggitt** [comment]
> It may have become disabled in 1.18 but in 1.16 you are unable to use the resource again

That depends on the configuration of the cluster the client is speaking to... administrators have the option to re-enable the deprecated APIs until 1.18.

> isn't the intent of https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1693-warnings to report when the resource is/will-be no longer served?

Yes, and 1.18+ is when the resource will definitely no longer be served. In 1.16 and 1.17, it could still be served if the cluster admin explicitly enabled those resources.

> We would like to be able to provide them with a message that informs them of the correct version when deprecated and removed.

We didn't think that hard about whether to indicate 1.16 or 1.18 as the removed version for these resources since either one referred to a release that was already past. If a client is directly introspecting the API types, I suppose it is better to err on the early side (and maybe the particular cluster they speak to would still have them available in 1.16/1.17)?

**Comment by joejulian** [comment]
If you warn a user that something's going to be removed in 1.22 and it stops working in 1.20 (even though they can re-enable it through flags) that's a surprise to the user that leads to a bad experience.

Any thoughts on how to address that UX concern?

**Comment by hickeyma** [comment]
If you look at it this way, not everyone updates to the latest version of Kubernetes and when users do they do usually in their own timeframes.

Some users may have been unaware of the deprecation and removal of APIs until 1.16 when there was a big change. I would think most users are also using the default cluster set-up and not the extension period for a removed API.

**Comment by joejulian** [comment]
Cross posted:

> If a client is directly introspecting the API types, I suppose it is better to err on the early side (and maybe the particular cluster they speak to would still have them available in 1.16/1.17)?

I agree and think this would handle my UX concern.
